module/zfs/arc.c

```diff
@@ -6051,6 +6068,10 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
 				acb->acb_noauth = noauth_read;
 				acb->acb_nobuf = no_buf;
 				acb->acb_zb = *zb;
+				/*
+				 * XXX couldn't there be a arc_read with
+				 * done == NULL and pio != NULL?
+				 */
```

Comment:
@grwilson  please comment on this XXX?

Reply:
Passing in a `pio` typically means that the caller wants some kind of interlocking of the I/Os that are issued and could be interested in propagating errors up the zio tree. Passing in a `done` func means that the caller wants to do something with the data.  So what would it mean to pass in a `pio` but a NULL `done`? Maybe a prefetch that cares about a failed I/O and wants to do something about it? For completeness, this seems like an overlooked case that could be solved by unconditionally creating the `acb`. Another option would be to add an `IMPLY(done == NULL, pio == NULL)` or something check like that. This way any code that were to add such a call would ensure the proper testing of any changes (I don't think we currently have a caller which passes this condition). If you do want to add a check, then it should also be added in the regular cache miss code path as well.